kafka-0.10.x/src/main/scala/monix/kafka/KafkaProducer.scala

```diff
@@ -57,28 +57,33 @@ object KafkaProducer {
   /** Builds a [[KafkaProducer]] instance. */
   def apply[K, V](config: KafkaProducerConfig, sc: Scheduler)(
     implicit K: Serializer[K],
-    V: Serializer[V]): KafkaProducer[K, V] =
-    new Implementation[K, V](config, sc)
-
-  private final class Implementation[K, V](config: KafkaProducerConfig, sc: Scheduler)(
-    implicit K: Serializer[K],
-    V: Serializer[V])
-      extends KafkaProducer[K, V] with StrictLogging {
-
-    private val isCanceled = Atomic(false)
-
-    // Gets initialized on the first `send`
-    private lazy val producerRef = {
-      logger.info(s"Kafka producer connecting to servers: ${config.bootstrapServers.mkString(",")}")
+    V: Serializer[V]): KafkaProducer[K, V] = {
+    lazy val producerRef: ApacheProducer[K, V] = {
       val keySerializer = K.create()
       val valueSerializer = V.create()
       val configJavaMap = config.toJavaMap
       keySerializer.configure(configJavaMap, true)
       valueSerializer.configure(configJavaMap, false)
       new ApacheKafkaProducer[K, V](configJavaMap, keySerializer, valueSerializer)
     }
+    new Implementation[K, V](config, sc, producerRef)
+  }
+
+  /** Builds a [[KafkaProducer]] instance with provided Apache Producer. */
+  def apply[K, V](config: KafkaProducerConfig, sc: Scheduler, producerRef: ApacheProducer[K, V])(
```

Comment:
I think `producerRef` should be taken lazily, or as a `Task` / `Coeval`.
The implementation can close `ApacheProducer` and that should be only done by the producer (creator) of the resource.

Reply:
Is this what you mean?
```scala
  /** Builds a [[KafkaProducer]] instance with provided Apache Producer. */
  def apply[K, V](config: KafkaProducerConfig, sc: Scheduler, producer: Coeval[ApacheProducer[K, V]])(
    implicit K: Serializer[K],
    V: Serializer[V]): KafkaProducer[K, V] = {
    lazy val producerRef = producer.value()
    new Implementation[K, V](config, sc, producerRef)
  }
```

Reply:
Yes, although `Implementation`  would have to use `Coeval` as well, otherwise it will initialize right away

Reply:
Ok, pushed a commit that uses Coeval producer.